core/commonMain/src/kotlinx/serialization/encoding/ReorderingCompositeEncoder.kt

```diff
@@ -0,0 +1,240 @@
+        val classEncoder = beginStructure(descriptor)
+        bufferedCalls.forEach { fieldToEncode ->
+            fieldToEncode?.encoder?.invoke(classEncoder)
+        }
+        classEncoder.endStructure(descriptor)
+    }
+
+    override fun encodeBooleanElement(descriptor: SerialDescriptor, index: Int, value: Boolean) {
+        bufferEncoding(descriptor, index) { encodeBooleanElement(descriptor, index, value) }
+    }
+
+    override fun encodeByteElement(descriptor: SerialDescriptor, index: Int, value: Byte) {
+        bufferEncoding(descriptor, index) { encodeByteElement(descriptor, index, value) }
+    }
+
+    override fun encodeCharElement(descriptor: SerialDescriptor, index: Int, value: Char) {
+        bufferEncoding(descriptor, index) { encodeCharElement(descriptor, index, value) }
+    }
+
+    override fun encodeDoubleElement(descriptor: SerialDescriptor, index: Int, value: Double) {
+        bufferEncoding(descriptor, index) { encodeDoubleElement(descriptor, index, value) }
+    }
+
+    override fun encodeFloatElement(descriptor: SerialDescriptor, index: Int, value: Float) {
+        bufferEncoding(descriptor, index) { encodeFloatElement(descriptor, index, value) }
+    }
+
+    override fun encodeIntElement(descriptor: SerialDescriptor, index: Int, value: Int) {
+        bufferEncoding(descriptor, index) { encodeIntElement(descriptor, index, value) }
+    }
+
+    override fun encodeLongElement(descriptor: SerialDescriptor, index: Int, value: Long) {
+        bufferEncoding(descriptor, index) { encodeLongElement(descriptor, index, value) }
+    }
+
+    override fun <T : Any> encodeNullableSerializableElement(
+        descriptor: SerialDescriptor,
+        index: Int,
+        serializer: SerializationStrategy<T>,
+        value: T?
+    ) {
+        bufferEncoding(descriptor, index) { encodeNullableSerializableElement(descriptor, index, serializer, value) }
+    }
+
+    override fun <T> encodeSerializableElement(
+        descriptor: SerialDescriptor,
+        index: Int,
+        serializer: SerializationStrategy<T>,
+        value: T
+    ) {
+        bufferEncoding(descriptor, index) { encodeSerializableElement(descriptor, index, serializer, value) }
+    }
+
+    override fun encodeShortElement(descriptor: SerialDescriptor, index: Int, value: Short) {
+        bufferEncoding(descriptor, index) { encodeShortElement(descriptor, index, value) }
+    }
+
+    override fun encodeStringElement(descriptor: SerialDescriptor, index: Int, value: String) {
+        bufferEncoding(descriptor, index) { encodeStringElement(descriptor, index, value) }
+    }
+
+    override fun encodeInlineElement(descriptor: SerialDescriptor, index: Int): Encoder {
+        return BufferingInlineEncoder(descriptor, index)
+    }
+
+    override fun shouldEncodeElementDefault(descriptor: SerialDescriptor, index: Int): Boolean {
+        callShouldEncodeElementDefault = true
+        return true
+    }
```

Comment:
I don't think you need/want this. The delegate/final encoder should be happy if it is never called even for the "default" value. As such `callShouldElementDefault` isn't needed. If it were to return false the field is just not written (and kept null/skipped). 

If you wish to have the encoder to force encoding defaults (rather than just delegating it to the composite) just have a property for that. `return forceEncodingDefaults || delegateEncoder.shouldEncodeElementDefault(descriptor, index)`. I'd be very surprised if there is any `CompositeEncoder` that is stateful in this function.

Reply:
My main idea was to have a composite encoder just reordering, that way if the user is doing custom logic in the shouldEncodeDefault method, then it will continue working exactly as a standard encoding.

Reply:
There is no requirement for `shouldEncodeDefault` to be called at all. And per its own definition, there is no guarantee that it is followed by the encoder for that same index (so if you want to keep this, you have to change it to an int with the (original) element index/a negative value for absent)

Reply:
Oh I get it, if it returns false, then the encoding method is not called so no buffering is done. I would just try to not add another parameter here. I think I'll try to : use the final composite encoder's shouldEncodeDefault directly, or buffer the call.

Reply:
Has been changed to `compositeEncoderDelegate.shouldEncodeDefault` and implicitly does not buffer the encoding.